### PR TITLE
Add a 'standalone_minimal' preset to build a minimal, static, OS independent, self-contained binaries of stdlib.

### PR DESCRIPTION
**Pull request by kubamracek**
This PR is the first step to be able to produce a new variant of the Swift Runtime and Standard Library — a minimal, static, self-contained build that is not tied to a particular operation system.

The building pieces are:
- A new "freestanding" platform, which doesn't imply any particular OS.
- A "min-size-release" build-script mode, which optimizes for size.
- Some CMake options that allow specifying how to actually build the stdlib (which SDK, archs, and target triple to use).
- A new preset that builds a "freestanding-x86_64" stdlib in this mode.

(Note that this is strictly for the runtime/stdlib, not for the compiler.)

### CODE REVIEW

File: utils/build-presets.ini [hunk 1]
```diff
@@ -2365,6 +2365,29 @@ mixin-preset=stdlib_DA_standalone,build
 test
 validation-test
 
+[preset: stdlib_MSR_standalone_minimal,build]
```
Comment (gottesmm) [inline utils/build-presets.ini:2368]: Can you put a comment here explaining what MSR is? Also, Can you change it to MR? Generally each letter stands for a different category. I would even consider S for size. Your thoughts?

File: utils/swift_build_support/swift_build_support/targets.py [hunk 2]
```diff
@@ -207,6 +210,7 @@ class StdlibDeploymentTarget(object):
         iOS, iOSSimulator,
         AppleTV, AppleTVSimulator,
         AppleWatch, AppleWatchSimulator,
+        Generic,
```
Comment (gottesmm) [inline utils/swift_build_support/swift_build_support/targets.py:213]: Can you add a comment here about what generic means?

File: utils/swift_build_support/swift_build_support/targets.py [hunk 3]
```diff
@@ -179,6 +179,9 @@ class StdlibDeploymentTarget(object):
                                          sdk_name="WATCHOS_SIMULATOR",
                                          is_simulator=True)
 
+    Generic = Platform("generic", archs=["i386", "x86_64", "armv7", "armv7s",
```
Comment (gottesmm) [inline utils/swift_build_support/swift_build_support/targets.py:182]: Can you add a comment here about what generic means?

File: utils/build-presets.ini [hunk 4]
```diff
@@ -2365,6 +2365,29 @@ mixin-preset=stdlib_DA_standalone,build
...
+enable-experimental-concurrency=0
+build-swift-dynamic-sdk-overlay=0
+build-swift-dynamic-stdlib=0
+build-swift-static-stdlib=1
+stdlib-deployment-targets=generic-x86_64
+extra-cmake-options=
+    -DSWIFT_SDKS:STRING=GENERIC
+    -DSWIFT_PRIMARY_VARIANT_SDK:STRING=GENERIC
+    -DSWIFT_PRIMARY_VARIANT_ARCH:STRING=x86_64
+    -DSWIFT_GENERIC_SDK:STRING=macosx
```
Comment (gottesmm) [inline utils/build-presets.ini:2387]: I would rather us not use extra-cmake-options. Are there build-script options that you can use here? If not, can we add them?

File: test/CMakeLists.txt [hunk 5]
```diff
@@ -102,6 +102,7 @@ function(get_test_dependencies SDK result_var_name)
      ("${SDK}" STREQUAL "IOS_SIMULATOR") OR
      ("${SDK}" STREQUAL "TVOS_SIMULATOR") OR
      ("${SDK}" STREQUAL "WATCHOS_SIMULATOR") OR
+     ("${SDK}" STREQUAL "GENERIC") OR
```
Comment (gottesmm) [inline test/CMakeLists.txt:105]: I hate this part of the code = (. [not your fault though].

File: cmake/modules/DarwinSDKs.cmake [hunk 6]
```diff
@@ -26,6 +26,23 @@ if(swift_build_osx)
...
+      "Which triple name (e.g. 'macosx11.0') to use when building the GENERIC stdlib")
+  set(SWIFT_GENERIC_ARCHS "" CACHE STRING
+      "Which architectures to build when building the GENERIC stdlib")
+  configure_sdk_darwin(
+      GENERIC "GENERIC" ""
+      "${SWIFT_GENERIC_SDK}" generic "${SWIFT_GENERIC_TRIPLE_NAME}" generic "${SWIFT_GENERIC_ARCHS}")
+  set(SWIFT_SDK_GENERIC_LIB_SUBDIR "generic")
+  configure_target_variant(GENERIC-DA "GENERIC Debug+Asserts"   GENERIC DA "Debug+Asserts")
+  configure_target_variant(GENERIC-RA "GENERIC Release+Asserts" GENERIC RA "Release+Asserts")
+  configure_target_variant(GENERIC-R  "GENERIC Release"         GENERIC R  "Release")
```
Comment (gottesmm) [inline cmake/modules/DarwinSDKs.cmake:43]: Do we need to add an S variant to this? I forgot what this configure target variant code does (I need to read it more).

File: cmake/modules/AddSwift.cmake [hunk 7]
```diff
@@ -131,7 +131,12 @@ function(_add_host_variant_c_compile_flags target)
 
   is_build_type_optimized("${CMAKE_BUILD_TYPE}" optimized)
   if(optimized)
-    target_compile_options(${target} PRIVATE -O2)
+    is_build_type_optimized_for_size("${CMAKE_BUILD_TYPE}" optimized_for_size)
+    if(optimized_for_size)
+      target_compile_options(${target} PRIVATE -Os)
+    else()
+      target_compile_options(${target} PRIVATE -O2)
```
Comment (gottesmm) [inline cmake/modules/AddSwift.cmake:138]: Another thing to look at is from LLVM IIRC there is a command line flag -O that we import. Can you just validate that this overrides that? (I believe this comes later on the command line so it overrides, but it would be good to check).

File: stdlib/cmake/modules/AddSwiftStdlib.cmake [hunk 8]
```diff
@@ -180,7 +180,12 @@ function(_add_target_variant_c_compile_flags)
 
   is_build_type_optimized("${CFLAGS_BUILD_TYPE}" optimized)
   if(optimized)
-    list(APPEND result "-O2")
+    is_build_type_optimized_for_size("${CFLAGS_BUILD_TYPE}" optimized_for_size)
+    if(optimized_for_size)
+      list(APPEND result "-Os")
+    else()
```
Comment (gottesmm) [inline stdlib/cmake/modules/AddSwiftStdlib.cmake:186]: This code repeating is sort of unfortunate. Maybe refactor it into a helper?

**Review by gottesmm — COMMENTED** [review]
Some thoughts/questions/etc.

File: utils/build-presets.ini (hunk 1 above)
Comment (kubamracek) [reply utils/build-presets.ini:2368]: > I would even consider S for size. Your thoughts?
Do you mean just "S"? That actually sounds the best to me.

File: cmake/modules/AddSwift.cmake (hunk 7 above)
Comment (kubamracek) [reply cmake/modules/AddSwift.cmake:138]: Yes, I checked, this is present later on the command lines.

File: stdlib/cmake/modules/AddSwiftStdlib.cmake (hunk 8 above)
Comment (kubamracek) [reply stdlib/cmake/modules/AddSwiftStdlib.cmake:186]: Something like `clang_optimize_flag_for_build_type()`?

File: utils/build-presets.ini (hunk 4 above)
Comment (kubamracek) [reply utils/build-presets.ini:2387]: Yes, but also let me explain here why each of these is needed:
- `SWIFT_SDKS=GENERIC` ... When building Darwin, the build insists to always build the macOS platform, SWIFT_SDKS will default to "GENERIC;OSX". This line prevents that.
- `SWIFT_PRIMARY_VARIANT_SDK`, `SWIFT_PRIMARY_VARIANT_ARCH` ... On Darwin, these default to macOS+x86_64, but if we're not building the macOS platform, a bunch of targets add dependencies to a non-existent target.
- `SWIFT_GENERIC_SDK` ... This configures the "generic" platform to select which SDK to actually use when building.

File: utils/build-presets.ini (hunk 1 above)
Comment (gottesmm) [reply utils/build-presets.ini:2368]: Yes

File: cmake/modules/AddSwift.cmake (hunk 7 above)
Comment (gottesmm) [reply cmake/modules/AddSwift.cmake:138]: Ok good. Thanks! I just wanted to be careful it didn't change.

File: stdlib/cmake/modules/AddSwiftStdlib.cmake (hunk 8 above)
Comment (gottesmm) [reply stdlib/cmake/modules/AddSwiftStdlib.cmake:186]: I would be fine with that.

File: cmake/modules/AddSwift.cmake (hunk 7 above)
Comment (compnerd) [reply cmake/modules/AddSwift.cmake:138]: Wait, I'm very confused why these are needed at all.  You are checking `CMAKE_BUILD_TYPE` - the flags for that should already be set properly (though it may do `-O3` instead of `-O2`).  What did I overlook here?  `CMAKE_<LANG>_FLAGS_${CMAKE_BUILD_TYPE}` is the right variable to use rather than hardcoding `-Os` and `-O2` - and it will ensure that things are handled properly for MSVC as well.

File: cmake/modules/SwiftUtils.cmake [hunk 9]
```diff
@@ -95,6 +95,18 @@ function(is_build_type_optimized build_type result_var_name)
   endif()
 endfunction()
 
+function(is_build_type_optimized_for_size build_type result_var_name)
```
Comment (compnerd) [inline cmake/modules/SwiftUtils.cmake:98]: Addressing the previous comment should probably obviate this.

File: cmake/modules/DarwinSDKs.cmake (hunk 6 above)
Comment (kubamracek) [reply cmake/modules/DarwinSDKs.cmake]: Add, but see the implementation of `configure_target_variant`:

```
# FIXME: this is not wired up with anything yet.
function(configure_target_variant prefix name sdk build_config lib_subdir)
```

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test

File: utils/build-script-impl [hunk 10]
```diff
@@ -223,6 +223,12 @@ KNOWN_SETTINGS=(
     report-statistics                             "0"               "set to 1 to generate compilation statistics files for swift libraries"
     sil-verify-all                                "0"               "If enabled, run the SIL verifier after each transform when building Swift files during this build process"
     stdlib-deployment-targets                     ""                "space-separated list of targets to configure the Swift standard library to be compiled or cross-compiled for"
+    swift-sdks                                    ""                "space-separated list of SDKs to build, or empty to auto-detect which SDKs are available"
```
Comment (gottesmm) [inline utils/build-script-impl:226]: Why did you add swift-sdks here. Are you still using this?

File: utils/build-presets.ini [hunk 11]
```diff
@@ -2365,6 +2365,30 @@ mixin-preset=stdlib_DA_standalone,build
 test
 validation-test
 
+[preset: stdlib_S_standalone_minimal,build]
+mixin-preset=stdlib_base_standalone
+
+build-subdir=stdlib_S_standalone
```
Comment (gottesmm) [inline utils/build-presets.ini:2371]: Some thoughts:

1. We probably want to have a couple of different presets here. Specifically, on CI we definitely want to compile with asserts enabled. At least in the compiler.
2. Can you split out the non-specific stuff into a base preset that is stdlib_S_standalone,build and have this inherit from that with the minimal stuff?

File: utils/build-script-impl (hunk 10 above)
Comment (kubamracek) [reply utils/build-script-impl:226]: Yes, I set it to avoid building for “OSX”.

File: utils/build-presets.ini (hunk 11 above)
Comment (kubamracek) [reply utils/build-presets.ini:2371]: Makes sense. Will do.

Comment (kubamracek) [reply utils/build-presets.ini:2371]: Ok, so I added separate 'S' and 'SA' presets that aren't specific, and then a mixin `mixin_stdlib_minimal` which drops certain features we don't want in the "minimal" variant. I plan to add more things into this mixin. Then I've renamed the actual preset to `stdlib_S_standalone_minimal_macho_x86_64`, because that's what it is: It's making a concrete build of this variant for x86_64+Mach-O. This way, someone can add an ELF concrete build if they want.

File: cmake/modules/AddSwift.cmake [hunk 12]
```diff
@@ -131,7 +131,8 @@ function(_add_host_variant_c_compile_flags target)
 
   is_build_type_optimized("${CMAKE_BUILD_TYPE}" optimized)
   if(optimized)
-    target_compile_options(${target} PRIVATE -O2)
+    clang_optimize_flag_for_build_type("${CMAKE_BUILD_TYPE}" optimize_flag)
+    target_compile_options(${target} PRIVATE "${optimize_flag}")
```
Comment (compnerd) [inline cmake/modules/AddSwift.cmake:135]: Why is this needed?  Why is the default not sufficient?  Why can `${CMAKE_CXX_FLAGS_${CMAKE_BUILD_TYPE}}` not sufficient?

File: cmake/modules/SwiftUtils.cmake [hunk 13]
```diff
@@ -95,6 +95,32 @@ function(is_build_type_optimized build_type result_var_name)
...
+         "${build_type}" STREQUAL "Release")
+    set("${result_var_name}" "-O2" PARENT_SCOPE)
+  elseif("${build_type}" STREQUAL "MinSizeRel")
+    set("${result_var_name}" "-Os" PARENT_SCOPE)
+  else()
+    message(FATAL_ERROR "Unknown build type: ${build_type}")
+  endif()
+endfunction()
+
+function(swift_optimize_flag_for_build_type build_type result_var_name)
```
Comment (compnerd) [inline cmake/modules/SwiftUtils.cmake:111]: Can you put this in `stdlib/cmake/modules` please?

File: cmake/modules/SwiftUtils.cmake [hunk 14]
```diff
@@ -95,6 +95,32 @@ function(is_build_type_optimized build_type result_var_name)
   endif()
 endfunction()
 
+function(clang_optimize_flag_for_build_type build_type result_var_name)
```
Comment (compnerd) [inline cmake/modules/SwiftUtils.cmake:98]: I don't think that we should have this function.  See the question about about why this is needed.  Also this doesn't take into account MSVC currently.  The content of this function should be equivalent to:

```
  set(${result_var_name} ${CMAKE_CXX_FLAGS_${CMAKE_BUILD_TYPE}} PARENT_SCOPE)
```

File: cmake/modules/AddSwift.cmake (hunk 12 above)
Comment (kubamracek) [reply cmake/modules/AddSwift.cmake:135]: So I think you're pointing out the right thing, but there's at least one behavior change if we did that: `CMAKE_CXX_FLAGS_RELEASE` is -O3, whereas we today do -O2. How about I send out a separate PR with to try to improve this bit here?

Comment (kubamracek) [reply cmake/modules/AddSwift.cmake:135]: (I don't want to change any observable behavior of any other preset with this PR.)

File: cmake/modules/SwiftUtils.cmake (hunk 14 above)
Comment (kubamracek) [reply cmake/modules/SwiftUtils.cmake:98]: Ok, undestood. However, I don't want to switch to `CMAKE_CXX_FLAGS_${CMAKE_BUILD_TYPE}` with this PR, as that's a change to -O3 from -O2 for some presets.

Would you prefer if I instead of introducing this function kept the original code that adds -O2 for optimized builds, and after it added another -Os (last flag wins) when CMAKEBUILD_TYPE is "MinSizeRel"?

Comment (compnerd) [reply cmake/modules/SwiftUtils.cmake:98]: Yeah, keeping the original would be preferable.

Comment (kubamracek) [reply cmake/modules/SwiftUtils.cmake:98]: https://github.com/apple/swift/pull/33388

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test macOS platform
